feat: reward one participant per inet group

bajtos opened the pull request:
Links:
- https://github.com/filecoin-station/spark/issues/29

bajtos commented:
@juliangruber I would like to get early feedback on the implementation - does it look right to you?

Existing tests are failing, and there are no new tests yet; those are TODO items I am aware of.

bajtos commented:
Dry run with partial data:

```
❯ node bin/dry-run.js 0x3113b83ccec38a18df936f31297de490485d7b2e 939 bafybeibebamtv747koblj4xxfy3owmj67oxygoj7mfaerftrzdaju4cdha bafybeiasipko2zpcllb3wxzptq6j7yp7jjftrtnfmssjs3reuobyjxldea bafybeihxmsl3pw2y6horqj5ut7vfujfew635znv6b4cnyxizwqzpprawmi
Evaluating round 939 of contract 0x3113b83ccec38a18df936f31297de490485d7b2e
==PREPROCESS==
PREPROCESS ROUND 939: Added measurements from CID bafybeibebamtv747koblj4xxfy3owmj67oxygoj7mfaerftrzdaju4cdha
{ total: 35, valid: 35 }
PREPROCESS ROUND 939: Added measurements from CID bafybeiasipko2zpcllb3wxzptq6j7yp7jjftrtnfmssjs3reuobyjxldea
{ total: 15, valid: 15 }
PREPROCESS ROUND 939: Added measurements from CID bafybeihxmsl3pw2y6horqj5ut7vfujfew635znv6b4cnyxizwqzpprawmi
{ total: 34, valid: 34 }
Fetched 84 measurements
==EVALUATE==
EVALUTE ROUND 939: Evaluated 84 measurements, found 82 honest entries.
{ OK: 82, INVALID_TASK: 0, DUPLICATE_SUBNET: 2 }
EVALUTE ROUND 939: Invoking IE.setScores(); number of participants: 2, total score: 999999999999999n
==EVALUATION RESULTS==
participants: [
  '0x000000000000000000000000000000000000dEaD',
  '0xf2D811C2aC3E8019c18FF3530777c072cc8a580A'
]
scores: [ 902439024390243n, 97560975609756n ]
==END OF RESULTS==
EVALUTE ROUND 939: IE.setScores() TX hash: 0x234
```

juliangruber commented on lib/evaluate.js:
```diff
@@ -89,19 +92,81 @@ export const evaluate = async ({
+    // 1. Hash the `finished_at` timestamp recorded by the server
+    // 2. Pick the measurement with the lowest hash value
+    // This relies on the fact that the hash function has a random distribution.
+    // We are also making the assumption that each measurement has a distinct `finished_at` field.
```
I guess even if `finished_at` values are redundant, the first measurement with the lowest hash will still win (measurement order is given), so the algorithm is still deterministic.

Why don't we reward the first measurement and instead the one with the lowest hash?

bajtos replied:
> Why don't we reward the first measurement

Sorting by `finished_at` alone would incentivise a race between SPARK nodes. The fastest node on the quickest internet connection would consistently win the reward.

By hashing the value, we get a different value-to-compare every round.

juliangruber commented:
> @juliangruber I would like to get early feedback on the implementation - does it look right to you?

Good idea to ask early 🙌

bajtos commented on lib/typings.d.ts:
```diff
@@ -21,3 +21,25 @@ export type RecordTelemetryFn = (
+  | 'OK'
+  | 'INVALID_TASK'
+  | 'NO_INET_GROUP'
+  | 'DUPLICATE_SUBNET'
```
Sounds good. I am concerned about data size, so I want to keep these keys short. I'll use `DUP_INET_GROUP`.

bajtos commented:
I added new telemetry fields to tell us how often participants win or loose the per-inet-group task reward. I fixed the failing tests, added tests for the new functionality and fixed bugs I discovered along the way.

This PR is done and ready for review.

bajtos commented:
Dry run of a partial round:

<details>
<summary>Full log - click to expand</summary>
<pre><code>
❯ node bin/dry-run.js 0x3113b83ccec38a18df936f31297de490485d7b2e 1256 bafybeiaztizd646qbmpkk5jvrp4rflzfwbmnqwirwlexdy6xuf5swaubia bafybeie2hpekq5govvezcruveiypakqppmwgmp6ajv2nffzhgbgwhldgsm bafybeibem7otp4uhzlhe474wgodqcucsljqffzy4k3npfbqhoj3b6hv5xq bafybeicf7i6kxvxvjxrqqbnqwhn6556lyfukjwci3siclyusv6vevai6am bafybeibzsueo2vh6acbbbart2uc25c2zgey2qdoubd5uqdafd3kqybcwva bafybeigoxmsgszbaqca4qhibpiwlunlztjz7u3xuwaebjti7rjttmapsle
Evaluating round 1256 of contract 0x3113b83ccec38a18df936f31297de490485d7b2e
==PREPROCESS==
PREPROCESS ROUND 1256: Added measurements from CID bafybeiaztizd646qbmpkk5jvrp4rflzfwbmnqwirwlexdy6xuf5swaubia
{ total: 19, valid: 19 }
PREPROCESS ROUND 1256: Added measurements from CID bafybeie2hpekq5govvezcruveiypakqppmwgmp6ajv2nffzhgbgwhldgsm
{ total: 47, valid: 47 }
PREPROCESS ROUND 1256: Added measurements from CID bafybeibem7otp4uhzlhe474wgodqcucsljqffzy4k3npfbqhoj3b6hv5xq
{ total: 31, valid: 31 }
PREPROCESS ROUND 1256: Added measurements from CID bafybeicf7i6kxvxvjxrqqbnqwhn6556lyfukjwci3siclyusv6vevai6am
{ total: 31, valid: 31 }
PREPROCESS ROUND 1256: Added measurements from CID bafybeibzsueo2vh6acbbbart2uc25c2zgey2qdoubd5uqdafd3kqybcwva
{ total: 37, valid: 37 }
PREPROCESS ROUND 1256: Added measurements from CID bafybeigoxmsgszbaqca4qhibpiwlunlztjz7u3xuwaebjti7rjttmapsle
{ total: 43, valid: 43 }
Fetched 208 measurements
==EVALUATE==
EVALUATE ROUND 1256: Evaluated 208 measurements, found 174 honest entries.
{
  OK: 174,
  INVALID_TASK: 0,
  NO_FINISHED_AT: 0,
  NO_INET_GROUP: 0,
  DUP_INET_GROUP: 34
}
EVALUATE ROUND 1256: Success rate of winning per-inet-group task reward:
{ min: 1, max: 1, mean: 1 }
EVALUATE ROUND 1256: Invoking IE.setScores(); number of participants: 27, total score: 999999999999989n
==EVALUATION RESULTS==
participants: [
  '0xC00bF99D112A85BcBdBDD9c1293EA55Af03bF9A2',
  '0x0bB7785dC278Fd5fC896B0CA1fc30e75484f8cF3',
  '0x7f2645e86dbF5fd6fCF613073bD795D0AA8546f5',
  '0x20b39BC37cEC6E03B4BF40ffd93bEF7445CbDd36',
  '0x67CbfC7b634ed14e77Cd61FfD24d911D5FF3F2B7',
  '0xC59095cD4051fb5227b7E0834BF0aCa04bD1Ea52',
  '0x18C742D8E19db9eb098E519eB2Fc02f0DEB6EAaA',
  '0x7A804Ef3d70a4683F8349D328C455Aa99dc11391',
  '0x77C04FEe4D440f89363Bf304b903C75fB5F0F9CD',
  '0x5E01D6f3F540D3E6Cbbf0Af03D6cF1133ec057F4',
  '0x000000000000000000000000000000000000dEaD',
  '0x4A7ECCBd710a0fD68F684739fAee1020bC09F395',
  '0xf2D811C2aC3E8019c18FF3530777c072cc8a580A',
  '0xe8F490895902f3b4010c7c808Bf56fA28E4b7126',
  '0xe27E98019E37d60dca0FCab245674250C53E79cC',
  '0x31276e69a720f537c02C45DD0c7e6F3b99e9F1d3',
  '0x9591b7ABf33A367D1860b346F7d3Ac3ba54CFCe9',
  '0x9CC725EaB99DD01320c8185A78FCDAF725431555',
  '0x1e2974e38E5a0515dcfFc6dBeE87929Bc36Db4e2',
  '0x6F3E06b1e0E0923339DcB08c3A2bD180dC3012Ec',
  '0xBbb3aD533771e4b016ff243B7023ed0b1c447a32',
  '0x328eF1442C1D502eE9db6a2A59C5378FDb9b73F4',
  '0xac20700326d29dE179Bfcd9B792b4773480895f0',
  '0x4cCF27bBEd39D65c0288c56D7305c537d15B08e3',
  '0x961C4d8d15F4372525Ae5640852BcaF5290532C3',
  '0x2fDbEdf4611b5725735b076b4B51e99349747591',
  '0x52FD73C40dF819abBec7fd188574b69dA1D89feA'
]
scores: [
   34482758620689n, 28735632183908n,
   45977011494252n, 40229885057471n,
  109195402298850n, 40229885057471n,
   28735632183908n, 34482758620689n,
   28735632183908n, 34482758620689n,
   97701149425287n, 34482758620689n,
   28735632183908n, 40229885057471n,
   40229885057471n, 40229885057471n,
   22988505747126n, 40229885057471n,
   51724137931034n, 28735632183908n,
   34482758620689n, 22988505747126n,
   22988505747126n, 17241379310344n,
   22988505747126n, 17241379310344n,
   11494252873563n
]
==END OF RESULTS==
EVALUATE ROUND 1256: IE.setScores() TX hash: 0x234
</code></pre>
</details>


The interesting bits:

```
EVALUATE ROUND 1256: Evaluated 208 measurements, found 174 honest entries.
{
  OK: 174,
  INVALID_TASK: 0,
  NO_FINISHED_AT: 0,
  NO_INET_GROUP: 0,
  DUP_INET_GROUP: 34
}
EVALUATE ROUND 1256: Success rate of winning per-inet-group task reward:
{ min: 1, max: 1, mean: 1 }
```

Even though we found some duplicate measurements in some inet groups, each node was rewarded for each task they completed. I assume nodes are sending duplicate measurements of the same task because of the low number of tasks to pick from.

bajtos commented:
@juliangruber hindsight please 🙏🏻

juliangruber reviewed (commented):
lgtm! 👏
